### PR TITLE
Add Force Monocolor rule

### PR DESCRIPTION
**Pull request by Distrib-ps**
PR for https://www.smogon.com/forums/threads/add-a-custom-rule-force-monocolor.3730674/

### CODE REVIEW

File: data/rulesets.ts [hunk 1]
```diff
@@ -489,6 +489,30 @@ export const Rulesets: {[k: string]: FormatData} = {
...
+		effectType: 'ValidatorRule',
+		name: 'Force Monocolor',
+		desc: `Forces all teams to have Pokémon of the same color. Usage: Force Monocolor = [Color], e.g. "Force Monocolor = Blue"`,
+		hasValue: true,
+		onValidateRule(value) {
+			const color = value.toLowerCase();
+			const validColors = ['red', 'blue', 'green', 'yellow', 'pink', 'brown', 'purple', 'gray', 'white', 'black'];
+			if (this.dex.gen <= 4) {
+				throw new Error(`The "Force Monocolor" rule is not supported in Generation ${this.dex.gen}.`);
+			}
```
[the remark this answers is not in the document]
Comment (Zarel) [reply data/rulesets.ts:502]: Like it's not in the data? You can always do `dex.mod('gen5').species.get(...).color`

[the remark this answers is not in the document]
Comment (TheoPonera) [reply data/rulesets.ts:502]: I'll look at that, thanks!

**Comment by KrisXV** [comment]
I don't know why we're adding this, there's really never been any demand for this outside of that suggestion and this is easily playable in challenges without this rule enforcement. A format for this will likely never happen either, so I don't think this is a rule worth investigating much.

**Comment by Distrib-ps** [comment]
> I don't know why we're adding this, there's really never been any demand for this outside of that suggestion and this is easily playable in challenges without this rule enforcement. A format for this will likely never happen either, so I don't think this is a rule worth investigating much.

I had a request for events on the Arcade room

File: data/rulesets.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Distrib-ps) [reply data/rulesets.ts]: If i use `this.dex.mod('gen5');` it works

File: data/rulesets.ts [hunk 2]
```diff
@@ -489,6 +489,30 @@ export const Rulesets: {[k: string]: FormatData} = {
...
+		hasValue: true,
+		onValidateRule(value) {
+			const color = value.toLowerCase();
+			const validColors = ['red', 'blue', 'green', 'yellow', 'pink', 'brown', 'purple', 'gray', 'white', 'black'];
+			if (this.dex.gen <= 4) {
+				this.dex = this.dex.mod('gen5');
+			}
+			if (!validColors.includes(color)) {
+				throw new Error(`Invalid color "${value}"`);
+			}
```
Comment (Karthik99999) [inline data/rulesets.ts:505]:
```suggestion
			const validColors = ['red', 'blue', 'green', 'yellow', 'pink', 'brown', 'purple', 'gray', 'white', 'black'];
			if (!validColors.includes(this.dex.toID(value))) {
				throw new Error(`Invalid color "${value}"`);
			}
```

File: data/rulesets.ts [hunk 3]
```diff
@@ -489,6 +489,30 @@ export const Rulesets: {[k: string]: FormatData} = {
...
+		},
+		onValidateSet(set) {
+			const ruleColor = this.ruleTable.valueRules.get('forcemonocolor')!.toLowerCase();
+			const species = this.dex.species.get(set.species);
+			const color = species.color.toLowerCase();
+			if (color !== ruleColor) {
+				return [`${set.species} must have the ${ruleColor} color.`];
+			}
```
Comment (Karthik99999) [inline data/rulesets.ts:513]:
```suggestion
			const color = this.toID(this.ruleTable.valueRules.get('forcemonocolor'));
			let dex = this.dex;
			if (dex.gen < 5) {
				dex = dex.forGen(5);
			}
			const species = dex.species.get(set.species);
			if (this.toID(species.color) !== color) {
				return [`${set.species} must be the color ${color}.`];
			}
```

**Comment by KrisXV** [comment]
something is coming up where this will be of use, so it's getting merged now. thanks!

File: data/rulesets.ts [hunk 4]
```diff
@@ -513,6 +513,29 @@ export const Rulesets: {[k: string]: FormatData} = {
...
 		},
 	},
+	forcemonocolor: {
+		effectType: 'ValidatorRule',
+		name: 'Force Monocolor',
+		desc: `Forces all teams to have Pok&eacute;mon of the same color. Usage: Force Monocolor = [Color], e.g. "Force Monocolor = Blue"`,
+		hasValue: true,
+		onValidateRule(value) {
+			const validColors = ["Black", "Blue", "Brown", "Gray", "Green", "Pink", "Purple", "Red", "White", "Yellow"];
+			if (!validColors.map(this.toID).includes(this.dex.toID(value))) {
```
Comment (KrisXV) [inline data/rulesets.ts:523]:
```suggestion
			if (!validColors.map(this.dex.toID).includes(this.dex.toID(value))) {
```

**Pull request merged**